Binary/Loader/FileSystemLoader.php

```diff
@@ -38,30 +38,19 @@ class FileSystemLoader implements LoaderInterface
     /**
      * @param MimeTypeGuesserInterface  $mimeGuesser
      * @param ExtensionGuesserInterface $extensionGuesser
-     * @param string[]                  $dataRoots
      * @param LocatorInterface          $locator
+     * @param string[]                  $rootPaths
      */
     public function __construct(
         MimeTypeGuesserInterface $mimeGuesser,
         ExtensionGuesserInterface $extensionGuesser,
-        $dataRoots
-        /* LocatorInterface $locator */
+        LocatorInterface $locator,
+        array $rootPaths = []
     ) {
         $this->mimeTypeGuesser = $mimeGuesser;
         $this->extensionGuesser = $extensionGuesser;
-
-        if (count($dataRoots) === 0) {
-            throw new InvalidArgumentException('One or more data root paths must be specified.');
-        }
-
-        if (func_num_args() >= 4 && false === ($this->locator = func_get_arg(3)) instanceof LocatorInterface) {
-            throw new \InvalidArgumentException(sprintf('Method %s() expects a LocatorInterface for the forth argument.', __METHOD__));
-        } elseif (func_num_args() < 4) {
-            @trigger_error(sprintf('Method %s() will have a forth `LocatorInterface $locator` argument in version 2.0. Not defining it is deprecated since version 1.7.2', __METHOD__), E_USER_DEPRECATED);
-            $this->locator = new FileSystemLocator();
-        }
-
-        $this->locator->setOptions(array('roots' => (array) $dataRoots));
+        $this->locator = $locator;
+        $this->locator->setOptions(['roots' => $rootPaths]);
```

Comment:
I have not found a straightforward solution to this, and based on my testing the solution will likely be quite large (only making this pull request even larger), so I don't forsee this issue being resolved by this pull request. See https://github.com/liip/LiipImagineBundle/pull/908#issuecomment-299718198 for additional details.

As for breaking changes (multiple times), I'm not too concerned about that; the whole point of having a development branch is to be able to _actively develop on it_. Those running the `1.x` branch should expect no backward incompatible changes, but that doesn't apply here. I run the `2.0` branch on a few personal projects, and while it's annoying to update it when breaking changes occur, that is a contract I agreed to when I decided to run an unreleased version. We're going to need to introduce a plethora of additional breaking changes to complete many of the TODO items for `2.0`, so let's focus on making the release amazing and not on making it easier for those running unreleased software. Right?